libretroshare/src/serialiser/rstypeserializer.h

```diff
@@ -439,176 +454,222 @@ struct RsTypeSerializer
 		}
 	}
 
-	/// std::vector<T>
+//============================================================================//
+//                       Sequence containers                                  //
+//============================================================================//
+
+	/** std::list is supported */ template <typename... Args>
+	struct is_sequence_container<std::list<Args...>>: std::true_type {};
+
+	/** std::set is supported */ template <typename... Args>
+	struct is_sequence_container<std::set<Args...>>: std::true_type {};
+
+	/**  std::vector is supported */ template <typename... Args>
+	struct is_sequence_container<std::vector<Args...>>: std::true_type {};
+
+
+	/// STL compatible sequence containers std::list, std::set, std::vector...
 	template<typename T>
-	static void serial_process( RsGenericSerializer::SerializeJob j,
-	                            RsGenericSerializer::SerializeContext& ctx,
-	                            std::vector<T>& v,
-	                            const std::string& memberName )
+	typename std::enable_if<is_sequence_container<T>::value>::type
+	static serial_process(
+	        RsGenericSerializer::SerializeJob j,
+	        RsGenericSerializer::SerializeContext& ctx,
+	        T& member, const std::string& memberName )
 	{
+		using el_t = typename T::value_type;
+
 		switch(j)
 		{
-		case RsGenericSerializer::SIZE_ESTIMATE:
-		{
-			ctx.mOffset += 4;
-			for(uint32_t i=0;i<v.size();++i)
-				serial_process(j,ctx,v[i],memberName);
-			break;
-		}
-		case RsGenericSerializer::DESERIALIZE:
-		{
-			uint32_t n=0;
-			serial_process(j,ctx,n,"temporary size");
-			v.resize(n);
-			for(uint32_t i=0;i<v.size();++i)
-				serial_process(j,ctx,v[i],memberName);
-			break;
-		}
+		case RsGenericSerializer::SIZE_ESTIMATE: // [[falltrough]]
 		case RsGenericSerializer::SERIALIZE:
 		{
-			uint32_t n=v.size();
-			serial_process(j,ctx,n,"temporary size");
-			for(uint32_t i=0; i<v.size(); ++i)
-				serial_process(j,ctx,v[i],memberName);
+			uint32_t aSize = member.size();
+			RS_SERIAL_PROCESS(aSize);
+			for(auto it = member.begin(); it != member.end(); ++it)
+			{
+				if(!ctx.mOk) break;
+				RS_SERIAL_PROCESS(*it);
+			}
 			break;
 		}
-		case RsGenericSerializer::PRINT:
+		case RsGenericSerializer::DESERIALIZE:
 		{
-			if(v.empty())
-				std::cerr << "  Empty vector \"" << memberName << "\""
-				          << std::endl;
-			else
-				std::cerr << "  Vector of " << v.size() << " elements: \""
-				          << memberName << "\"" << std::endl;
+			uint32_t elCount = 0;
+			RS_SERIAL_PROCESS(elCount);
+			if(!ctx.mOk) break;
+
+			/* This check is not perfect but will catch most pathological cases.
+			 * Avoid multiplying by sizeof(el_t) as it is not a good exitimation
+			 * of the actual serialized size, depending on the elements
+			 * structure and on the so it would raises many false positives. */
+			if(elCount > ctx.mSize - ctx.mOffset)
```

Comment:
since all these things are uint32_t, it's much safer to do 
   if(elCount + ctx.mOffset > ctx.mSize)
debugging bugs caused by overloading of uint32_t is pretty painful.

Reply:
While we control both `ctx.mSize` and `ctx.mOffset` and it i always true that `ctx.mOffset <= ctx.mSize` so  `ctx.mSize - ctx.mOffset >= 0` and then it is safe to do `ctx.mSize - ctx.mOffset`.  Instead we don't have control over `elCount` which is sent by the peer which could be bugged and send us a big value that would cause `(elCount + ctx.mOffset) > std::numeric_limits<uint32_t>::max()` that would cause an undetectable overflow

Reply:
eventually we can put in te code some runtime `assert(ctx.mOffset <= ctx.mSize)` but that should be not necessary as we control those values

Reply:
First if you put an assert you offer the possibility to crash friend's nodes and even distant nodes by sending crafted items.
Then, everywhere else in the code, we pay attention to not computing the difference between unsigned ints because some compilers do not handle this correctly in all cases.
Finally, moving one term the the left part is not a terrible change to do as compared to what it can prevent. 
So I really really do not understand your reaction here.

Reply:
also when you say "we have control on...", it's not a valid argument: bugs happen. And their effect should be limited. That is why we make methods bullet-proof on an individual basis. This is what makes a low level lib such as libretroshare stable.

Reply:
The peer have no way to craft `ctx.mSize` and `ctx.mOffset` while it can craft `elCount` because of that it is safer doing arithmetic operations on `ctx.mSize` and `ctx.mOffset` while it is not safe to do arithmetic operations on `elCount`. The compiler handle unsigned integers operation perfectly provided that the result of operation is > 0, which is the case. while the runtime have the same exact problem if the result of the operation is < 0 or > std::numeric_limits<uint32_t>::max(). The problem is not if we use + or - the problem is if the results are outside type limits.